From eb8cdb61e66d28b0f9b90ec64a321beb4b511927 Mon Sep 17 00:00:00 2001
From: Vivek Goyal <vgoyal@redhat.com>
Date: Mon, 8 Sep 2014 20:11:49 -0400
Subject: [block] cfq-iosched: Add comments on update timing of weight

Message-id: <1410207109-30525-3-git-send-email-vgoyal@redhat.com>
Patchwork-id: 92484
O-Subject: [RHEL7.1 PATCH 2/2] cfq-iosched: Add comments on update timing of weight
Bugzilla: 1116126
RH-Acked-by: Jeff Moyer <jmoyer@redhat.com>

BZ:1116126

commit 7b5af5cffce569298b1d03af1ddf1dc43570ab42
Author: Toshiaki Makita <makita.toshiaki@lab.ntt.co.jp>
Date: Thu, 28 Aug 2014 17:14:58 +0900

Explain that weight has to be updated on activation.
This complements previous fix e15693ef18e1 ("cfq-iosched: Fix wrong
children_weight calculation").

Signed-off-by: Toshiaki Makita <makita.toshiaki@lab.ntt.co.jp>
Signed-off-by: Jens Axboe <axboe@fb.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/block/cfq-iosched.c b/block/cfq-iosched.c
index 0d208a9..78be7af 100644
--- a/block/cfq-iosched.c
+++ b/block/cfq-iosched.c
@@ -1272,6 +1272,9 @@ __cfq_group_service_tree_add(struct cfq_rb_root *st, struct cfq_group *cfqg)
  rb_insert_color(&cfqg->rb_node, &st->rb);
 }
 
+/*
+ * This has to be called only on activation of cfqg
+ */
 static void
 cfq_update_group_weight(struct cfq_group *cfqg)
 {
@@ -1303,6 +1306,11 @@ cfq_group_service_tree_add(struct cfq_rb_root *st, struct cfq_group *cfqg)
  /* add to the service tree */
  BUG_ON(!RB_EMPTY_NODE(&cfqg->rb_node));
 
+ /*
+  * Update leaf_weight.  We cannot update weight at this point
+  * because cfqg might already have been activated and is
+  * contributing its current weight to the parent's child_weight.
+  */
  cfq_update_group_leaf_weight(cfqg);
  __cfq_group_service_tree_add(st, cfqg);
 
-- 
1.7.1